### PR TITLE
New template interface, default styles, and mobile template

### PR DESCRIPTION
**Pull request by mibrah42**
## What are we adding in this PR

- [x] New template interface that includes a configuration object.

```Typescript
interface Template {
 config: {
   parentID: string;
   parentName: string;
   id: string;
   name: string;
   textDirection: 'left' | 'right';
   language: string;
   device: 'desktop' | 'mobile';
 };
 // Translations are not part of the MVP, but will be added later like so.
 translations: {
  [languageCode: string]: {
     [widgetId: string]: string;
  };
 };
 widgets: {
   [id: string]: {
     id: string;
     editable?: boolean;
     widgetRef?: HTMLElement;
     children: string[];
     uniqueAttributes: { [key: string]: string };
     style: { [key: string]: string };
   };
 };
}
```

- [x] Default styling added to all widgets on creation.
       - Reduces JSON size.

- [x] Added `ui-sidemenu` widget which collapses on button press.

- [x] Added left drawer template.
image

- [x] Added device filters for selecting templates.
Screenshot 2020-07-28 at 6 41 49 PM

### CODE REVIEW

File: app-creator/client/package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
   "author": "Earth Engine Team",
   "license": "BSD-3-Clause",
   "dependencies": {
+    "@cwmr/paper-chip": "^3.0.0",
```
Comment (sufyanAbbasi) [inline app-creator/client/package.json:33]: Use this one instead: https://github.com/bendavis78/paper-chip

File: app-creator/client/src/redux/reducer.ts [hunk 2]
```diff
@@ -217,12 +224,29 @@ export const reducer: Reducer<AppCreatorStore, AppCreatorAction | AnyAction> = (
...
 };
 
+/**
+ * Adds default styles to all widgets.
+ */
+function addDefaultStyles(template: { [key: string]: any }) {
+  for (const id in template.widgets) {
+    template.widgets[id].style = {
+      ...DEFAULT_STYLES,
+      ...template.widgets[id].style,
```
Comment (sufyanAbbasi) [inline app-creator/client/src/redux/reducer.ts]: Though elegant, object spread is a ECMAScript 2018 feature and I don't think we're doing any compatibility transpiling. I'd suggest using Object.assign instead.

File: app-creator/client/src/utils/template-generation.ts [hunk 3]
```diff
@@ -105,16 +108,22 @@ export function getWidgetElement({
       element = wrapper;
       break;
     case WidgetType.panel:
+    case WidgetType.sidemenu:
       if ((element as Panel).editable) {
         (element as Panel).editable = editable ?? false;
```
Comment (sufyanAbbasi) [inline app-creator/client/src/utils/template-generation.ts]: I think the nullish coalescing doesn't afford any new logic here, since empty string and 0 are still falsey. More idiomatically is: `.editable = !!editable`. Also below.

File: app-creator/client/src/redux/types/enums.ts [hunk 4]
```diff
@@ -40,4 +41,5 @@ export enum EventType {
 export enum DeviceType {
   desktop = 'desktop',
   mobile = 'mobile',
+  all = 'all',
```
Comment (sufyanAbbasi) [inline app-creator/client/src/redux/types/enums.ts:45]: I totally missed this earlier, but enums, since they're constants, are better expressed in all caps. For now, I would add this as a TODO for cleaning up later.

File: app-creator/client/src/widgets/app-root.ts [hunk 5]
```diff
@@ -223,6 +262,34 @@ export class AppRoot extends LitElement {
       ? html`<paper-progress indeterminate></paper-progress>`
       : nothing;
 
+    const sortingChips = html`
+      <div id="chips-container">
+        <paper-chip
+          selectable
+          class="${this.deviceFilter === DeviceType.all
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/app-root.ts]: `classMap` is more succinct here: https://lit-element.polymer-project.org/guide/styles#classmap

File: app-creator/client/src/widgets/app-root.ts [hunk 6]
```diff
@@ -223,6 +262,34 @@ export class AppRoot extends LitElement {
...
+            ? 'selected-paper-chip'
+            : ''}"
+          @click=${() => handleDeviceFilters.call(this, DeviceType.all)}
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/app-root.ts]: I believe you don't need to bind this since Lit Element does it for you: https://lit-element.polymer-project.org/guide/events#using-this-in-event-handlers

File: app-creator/client/src/widgets/app-root.ts [hunk 7]
```diff
@@ -223,6 +262,34 @@ export class AppRoot extends LitElement {
...
+          >Web</paper-chip
+        ><paper-chip
+          selectable
+          class="${this.deviceFilter === DeviceType.mobile
+            ? 'selected-paper-chip'
+            : ''}"
+          @click=${() => handleDeviceFilters.call(this, DeviceType.mobile)}
+          >Mobile</paper-chip
+        >
+      </div>
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/app-root.ts]: Since we're just repeating the same component, I bet this can be more succinctly expressed as a map: https://lit-element.polymer-project.org/guide/templates#loops

File: app-creator/client/src/widgets/draggable-widget/draggable-widget.ts [hunk 8]
```diff
@@ -91,8 +91,14 @@ export class DraggableWidget extends LitElement {
...
 
-    if (type === WidgetType.panel) {
-      const dropzone = editingWidget.querySelector('dropzone-widget');
+    if (type === WidgetType.panel || type === WidgetType.sidemenu) {
+      let dropzone = editingWidget.querySelector('dropzone-widget');
+
+      if (dropzone == null) {
+        dropzone = editingWidget
+          .querySelector('slot')
+          ?.assignedNodes()[0] as Dropzone;
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/draggable-widget/draggable-widget.ts]: TIL: I've been using slot all wrong... This is awesome!

Question: This does not seem robust to if we decide to refactor the ordering of the dropzone component. Can we be more explicit about finding the dropzone widget in the slot?

File: app-creator/client/src/widgets/story-board/story-board.ts [hunk 9]
```diff
@@ -118,61 +129,28 @@ export class Storyboard extends connect(store)(LitElement) {
...
-            switchDeviceViewTab.call(this, DeviceType.mobile)}
-            ><iron-icon icon="hardware:smartphone"></iron-icon></paper-icon-button
-          ></paper-tab>
-        </paper-tabs>
-
-        <paper-card id="storyboard" style=${styleMap(styles)}></paper-card>
+        <paper-card
+          id="storyboard"
+          style=${styleMap(styles)}
+          class="storyboard ${isMobile ? 'mobile-storyboard' : ''}"
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/story-board/story-board.ts]: `classMap` is useful here too

File: app-creator/client/src/widgets/story-board/story-board.ts [hunk 10]
```diff
@@ -95,11 +104,13 @@ export class Storyboard extends connect(store)(LitElement) {
   `;
 
   stateChanged(state: AppCreatorStore) {
+    const template = state.template;
     if (
-      state.template.id !== this.templateID ||
-      state.eventType == EventType.importing
+      (template.hasOwnProperty('config') &&
+        template.config.id !== this.templateID) ||
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/story-board/story-board.ts]: I see, so you use the ID param on the template itself to determine whether we need to regenerate the template components. Similarly, we can achieve the same result by setting the ID directly on the redux state and keeping the local templateID cache to check against to avoid having to store the ID on the template.

In fact, I believe this mechanic can be used to completely factor out the `importing` `EventType` (best to remove as much state as possible), since we can check that the ID has changed to trigger any cloning logic etc. WDYT?

File: app-creator/client/src/widgets/templates-tab/templates-tab.ts [hunk 11]
```diff
@@ -85,15 +113,22 @@ export class TemplatesTab extends connect(store)(LitElement) {
...
+        name.toLowerCase().includes(lowerCasedQuery);
+
+      let matchesDevice = true;
+      if (deviceFilter !== DeviceType.all) {
+        matchesDevice = device === deviceFilter;
+      }
+
+      return matchesQuery && matchesDevice;
     });
   }
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/templates-tab/templates-tab.ts]: I think we can turn these 10 lines into one, succinct boolean expression (and include comments for better readability):
```js
templateCards.filter(({ id, name, device }) => 
  // Matches either the ID or name.
  query.match(new RegExp(`${id}|${name}`, 'i')) &&
  // And matches the device type (or all).
  (deviceFilter === DeviceType.all || device === deviceFilter)
);
```

File: app-creator/client/src/widgets/templates-tab/templates-tab.ts [hunk 12]
```diff
@@ -132,12 +178,41 @@ export class TemplatesTab extends connect(store)(LitElement) {
       ></empty-notice>
     `;
 
+    const sortingChips = html`
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/templates-tab/templates-tab.ts:176]: Same refactoring from above.

File: app-creator/client/src/widgets/ui-panel/ui-panel.ts [hunk 13]
```diff
@@ -125,18 +112,36 @@ export class Panel extends LitElement {
     // Remove highlight from currently selected widget (if any).
     DraggableWidget.removeEditingWidgetHighlight();
 
-    const dropzone = this.querySelector(
-      'dropzone-widget'
-    )?.shadowRoot?.querySelector(`#${CONTAINER_ID}`);
+    let dropzone = this.querySelector('dropzone-widget') as Dropzone;
+
+    if (dropzone == null) {
+      dropzone = this.querySelector('slot')?.assignedNodes()[0] as Dropzone;
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-panel/ui-panel.ts]: Same as above.

File: app-creator/client/src/widgets/ui-panel/ui-panel.ts [hunk 14]
```diff
@@ -125,18 +112,36 @@ export class Panel extends LitElement {
...
     store.dispatch(setEditingWidget(this));
   }
 
+  render() {
+    const { isRaised, layout, padded } = this;
+
+    return html`
+      <div
+        id="container"
+        class="${layout} ${isRaised ? 'raised' : ''} ${padded ? 'padded' : ''}"
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-panel/ui-panel.ts]: `classMap` here too

File: app-creator/client/src/widgets/templates-tab/templates-tab.ts (hunk 11 above)
Comment (sufyanAbbasi) [reply app-creator/client/src/widgets/templates-tab/templates-tab.ts]: Also, add a test to cover this logic.

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    .row {
+      flex-direction: row;
+    }
+  `;
+
+  /**
+   * Sets the flex layout of child widgets.
+   * Options available are 'column' and 'row'.
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts:66]: Make this an enum.

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    if (this.panel != null) {
+      this.panel.id = this.id;
+    }
+  }
+
+  toggleMenu(e: Event) {
+    e.stopPropagation();
+
+    const { panel } = this;
+    if (panel == null) {
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts]: `if (!panel)` covers `undefined` as well, though I can't imagine when or how `this` can be undefined...

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+      panel.style.width = '80%';
+    }
+  }
+
+  setStyle(style: { [key: string]: string }) {
+    const filteredStyles = new Set(['position', 'top', 'left', 'width']);
+    if (this.panel != null) {
+      for (const attribute in style) {
+        if (!filteredStyles.has(attribute)) {
+          console.log({ attribute });
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts]: Remove `console.log`.

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+    switch (key) {
+      case 'layout':
+        this.layout = value;
+        return;
+      case 'hasDropzone':
+        this.hasDropzone = value === 'true';
+        return;
+    }
+
+    this.requestUpdate();
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts:132]: Is it intentional that `this.requestUpdate` is not called if the key is `layout` or `hasDropzone`?  If so, it deserves a comment.

**Review by sufyanAbbasi — CHANGES_REQUESTED** [review]
Nice job! I left some suggestions for some major refactoring and some small nits. Let me know if you have any questions!

File: app-creator/client/src/widgets/app-root.ts (hunk 6 above)
Comment (mibrah42) [reply app-creator/client/src/widgets/app-root.ts]: Since we are passing the device type as an argument, I needed to wrap it with an anonymous function. The context `this` is only available in the anonymous function and not in the handleDeviceFilters function and thus I needed to bind it. Alternatively I could do this:

```Typescript
() => {
  this.deviceFilter = device;
}
```

File: app-creator/client/src/widgets/draggable-widget/draggable-widget.ts (hunk 8 above)
Comment (mibrah42) [reply app-creator/client/src/widgets/draggable-widget/draggable-widget.ts]: Good catch. Here is what we could do:
```Typescript
dropzone = editingWidget
          .querySelector('slot')
          ?.assignedNodes()
          .find((node) => node.nodeName === 'DROPZONE-WIDGET') as Dropzone;
```

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts (hunk 16 above)
Comment (mibrah42) [reply app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts]: Double equals`==` cover undefined as well, so if panel is null or undefined, it will return true. If we had `===` instead then it would not cover undefined. 

I use `if (panel == null)` instead of `if (!panel)` just to be more explicit of what we are expecting panel to be. It might be unnecessary for this case.

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts (hunk 18 above)
Comment (mibrah42) [reply app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts:132]: should've been a break statement

File: app-creator/client/src/widgets/templates-tab/templates-tab.ts (hunk 11 above)
Comment (mibrah42) [reply app-creator/client/src/widgets/templates-tab/templates-tab.ts]: That query will only match if the user searched for the exact name of the template. We want to match any substring. For example, we want to match `left` with `Left Side Panel`. 

We would need a regex like so:
```Typescript
    const queryMatch =
        id.match(new RegExp(`.*${lowerCasedQuery}.*`, 'i')) ||
        name.match(new RegExp(`.*${lowerCasedQuery}.*`, 'i'));
```

File: app-creator/client/package.json (hunk 1 above)
Comment (sufyanAbbasi) [reply app-creator/client/package.json:33]: Let's go ahead and do this in a follow up PR since it will be non-trivial (no npm pacakge for it).

File: app-creator/client/src/widgets/templates-tab/test/templates-tab_test.ts [hunk 19]
```diff
@@ -14,4 +16,56 @@ suite('templates-tab', () => {
...
   });
+
+  suite('filter template cards', () => {
+    test('Empty query from all devices', () => {
+      const result = TemplatesTab.filterTemplates(
+        templatesStub,
+        '',
+        DeviceType.all
+      );
+      expect(result.length).to.equal(templatesStub.length);
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/templates-tab/test/templates-tab_test.ts:21]: Thanks for adding these tests, they are great! I would recommend renaming this tests "Empty query for all device types" and enumerate the two other `DeviceType`s for completeness.

File: app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+  toggleMenu(e: Event) {
+    e.stopPropagation();
+
+    const { panel } = this;
+    if (!panel) {
+      return;
+    }
+
+    if (panel.style.width === '80%' || panel.style.width === '') {
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/ui-sidemenu/ui-sidemenu.ts]: Your toggler will get stuck if for some reason, the panel width is changed manually. Instead, use a regex to check for zero width and a ternary.

```js
const isZeroWidth = panel.style.width.match(/^\s*0(px|%)?/);
panel.style.width = isZeroWidth ? '80%' : '0%';
```

File: app-creator/client/src/widgets/templates-tab/templates-tab.ts [hunk 21]
```diff
@@ -85,15 +115,22 @@ export class TemplatesTab extends connect(store)(LitElement) {
...
   static filterTemplates(
-    templateCards: { id: string; name: string; markup: TemplateResult }[],
-    query: string
+    templateCards: TemplatesTabItem[],
+    query: string,
+    deviceFilter: DeviceType
   ): Array<TemplatesTabItem> {
-    return templateCards.filter(({ id, name }) => {
+    return templateCards.filter(({ id, name, device }) => {
       const lowerCasedQuery = query.toLowerCase();
```
Comment (sufyanAbbasi) [inline app-creator/client/src/widgets/templates-tab/templates-tab.ts]: The 'i' flag added to the `RegExp` makes the search case insensitive, so you don't have to lowercase the query.
You also don't need the `.*` since you didn't include the start and end characters `^` and `$`. Was there a particular case that was failing without it?

**Review by sufyanAbbasi — APPROVED** [review]
Thanks! I added a comment to fix up the search regex (message me if you have questions about it) and to add a few more test cases. Since they are fairly trivial, I will go ahead and approve now.
